zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs

```diff
@@ -0,0 +1,302 @@
+        //
+        // if (a > 0 && b > 0) || (a < 0 && b < 0):
+        //      a < b -> (a_hi < b_hi) ? 1 : (a_hi == b_hi) * (a_lo < b_lo)
+        // for example, for 8bit signed int, -1 is 0xff and -2 is 0xfe,
+        //     -2 < -1 and 0xfe < 0xff
+        let a_lt_b = select::expr(a_lt_b_hi, 1.expr(), a_eq_b_hi * a_lt_b_lo);
+
+        // Add a trivial selector: if only a or only b is negative we have the
+        // result.
+        // result = if a < 0 && b > 0, slt = 1.
+        // result = if b < 0 && a > 0, slt = 0.
+        let a_negative =
+            (1.expr() - sign_check_a_lt.expr()) * sign_check_b_lt.expr();
+        let b_negative =
+            (1.expr() - sign_check_b_lt.expr()) * sign_check_a_lt.expr();
+        let result = select::expr(
+            a_negative,
+            1.expr(),
+            select::expr(b_negative, 0.expr(), a_lt_b),
+        );
+
+        // Pop a and b from the stack, push the result on the stack.
+        cb.stack_pop(select::expr(is_sgt.expr(), b.expr(), a.expr()));
+        cb.stack_pop(select::expr(is_sgt.expr(), a.expr(), b.expr()));
+        cb.stack_push(result);
+
+        // The read-write counter changes by three since we're reading two words
+        // from stack and writing one. The program counter shifts only by one
+        // and the since the stack now has one less word, the stack pointer also
+        // shifts by one.
+        let step_state_transition = StepStateTransition {
+            rw_counter: Transition::Delta(3.expr()),
+            program_counter: Transition::Delta(1.expr()),
+            stack_pointer: Transition::Delta(1.expr()),
+            ..Default::default()
+        };
+        let same_context = SameContextGadget::construct(
+            cb,
+            opcode,
+            step_state_transition,
+            None,
+        );
+
+        Self {
+            same_context,
+            a,
+            b,
+            sign_check_a,
+            sign_check_b,
+            lt_lo,
+            comparison_hi,
+            is_sgt,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _transaction: &Transaction<F>,
+        _call: &Call<F>,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let opcode = step.opcode.unwrap();
+
+        // SLT opcode is the default check in the SCMP gadget. Swap rw for SGT.
+        self.is_sgt.assign(
+            region,
+            offset,
+            F::from(opcode.as_u8() as u64),
+            F::from(OpcodeId::SGT.as_u8() as u64),
+        )?;
+        let indices = if opcode == OpcodeId::SGT {
+            [step.rw_indices[1], step.rw_indices[0]]
+        } else {
+            [step.rw_indices[0], step.rw_indices[1]]
+        };
+        let [a, b] =
+            indices.map(|idx| block.rws[idx].stack_value().to_le_bytes());
```

Comment:
Just want to understand why we use `to_le_bytes` here?

Reply:
The RW table still uses little-endianness. The initial test failures (lookup errors) were in fact due to using `to_be_bytes()`.